snakemake/executors/google_lifesciences.py

```diff
@@ -897,24 +899,27 @@ def _retry_request(self, request, timeout=2, attempts=3):
         attempts: remaining attempts, throw error when hit 0
         """
         import googleapiclient
-
+        import google.auth
+        credentials, project_id = google.auth.default(scopes=["https://www.googleapis.com/auth/cloud-platform"])
```

Comment:
@vsoch do we authenticate this way in the other places as well? Or can there maybe be an auth object for the entire life science executor object instead creating a special one here in the retry?

Reply:
We go about it different ways, but I suspect under the hood since both are using the default application credentials (exported to the environment as `GOOGLE_APPLICATION_CREDENTIALS`) the methods are similar. We use the discovery API clients to authenticate clients that are attached to the entire class, e.g., here https://github.com/snakemake/snakemake/blob/417f40deb06a01d836c4ec632fc3b8f67672a403/snakemake/executors/google_lifesciences.py#L136-L165. And if you look into the source code, the googleapiclient discovery has that same "google.auth.default" function [here](https://github.com/googleapis/google-api-python-client/blob/9639c0709897ab4f56e4fb0e0fe78f2c339e59fe/googleapiclient/_auth.py#L51-L66) which will call that same auth function via [here](https://github.com/googleapis/google-api-python-client/blob/9639c0709897ab4f56e4fb0e0fe78f2c339e59fe/googleapiclient/discovery.py#L575-L579). Notably build takes an http object [here](https://github.com/googleapis/google-api-python-client/blob/9639c0709897ab4f56e4fb0e0fe78f2c339e59fe/googleapiclient/discovery.py#L179) and given that we don't provide one and it's generated,I think it might be saved on the returned service objects [here](https://github.com/googleapis/google-api-python-client/blob/9639c0709897ab4f56e4fb0e0fe78f2c339e59fe/googleapiclient/discovery.py#L1345) so I think what I would do is compare the object there (likely directly or its credentials) to the one created above. If they are comparable, arguably we could grab this object from one of the service objects generated from the discovery build API. The discovery build APIs are going to be different to have scopes for their respective endpoints, vs the one above which is all of google cloud. We could also try the reverse - creating one credential object and then seeing if we can pass to the services, although we'd want to double check how the scopes line up. And if neither of those ideas work, we could minimally have this call in the same function so they are in the same place!

Reply:
The request object is generated ultimately _from_ one of those services - either directly or as a result of doing like pipelines.run() so I'm curious why the request object isn't coming with its own http already (or maybe some of them are but not consistently?) E.g., pipelines.run() here is the one that was originally giving us trouble: https://github.com/snakemake/snakemake/blob/bf5fdfe2a0547bc20873ee4afcbd438928368c02/snakemake/executors/google_lifesciences.py#L831 and maybe @moschetti knows what using these endpoints and generated from a discovery endpoint, e.g, [here](https://cloud.google.com/life-sciences/docs/reference/rest/v2beta/projects.locations.pipelines/run) doesn't generate a request with a comparable http object that we are generating later in retry request?

Reply:
In my first attempt I tried passing the credentials created here: https://github.com/snakemake/snakemake/blob/417f40deb06a01d836c4ec632fc3b8f67672a403/snakemake/executors/google_lifesciences.py#L150
To `google_auth_httplib2.AuthorizedHttp` in `_retry_request`, but it did not like those credentials giving this error: `AttributeError: '_JWTAccessCredentials' object has no attribute 'before_request'` which led me to this [issue](https://github.com/googleapis/google-cloud-python/issues/3281), where I found to use `google.auth` for the credentials.